### PR TITLE
Box Wars Ep. IV: A New Box

### PR DESCRIPTION
**Pull request by chmeliik**
Adventures of The Box on openshift, part I.

To deploy this version of The Box, one currently needs to:
* set up an openshift cluster (with the ability to login as system:admin)
* install python-cryptography (or allow the generate-certs playbook to install it for them and run it with --ask-become-pass)
* run the generate-certs.yaml playbook (only once per year, technically)
* run the start-koji.yaml playbook
* pray that it works (it usually does)

TODO:
- [x] clean up config files
- [x] clean up ansible stuff (make it more role-ish I guess, or at least have a dedicated variables file)
- [x] check if the https:// version of koji web is working and why it isn't (it is :open_mouth: )
- [x] install our components from git
- [x] use a PV for the /mnt/koji/ directory, mount it on the builder as well
- [x] replicate koji setup somewhere (on the builder maybe)
- [x] update the README

TODO, but will not be part of this PR:
* a playbook that does openshift cluster setup
* the rest of the containerbuild machinery (buildroot, registry, namespaces)
* RHEL :heart_decoration:

### CODE REVIEW

File: client/bin/setup.sh [hunk 1]
```diff
@@ -1,66 +0,0 @@
...
-
-# TODO: Just create directly in DB?
-koji add-tag build --arches=x86_64
-koji add-tag dest --arches=x86_64
-koji add-target candidate build dest
-
-koji add-pkg dest osbs-buildroot-docker --owner kojiadmin
-koji add-pkg dest docker-hello-world --owner kojiadmin
-
-koji call addBType operator-manifests
```
Comment (MartinBasti) [inline client/bin/setup.sh]: Where are these koji settings replicated?

**Comment by MartinBasti** [comment]
Thank you, good job! It looks very nice (I haven't tried it yet but looking forward to deploying it).

I'd prefer usage of k8s module to setup application in OCP, but this is fine for starter (I prefer separate PR for that).

> run the generate-certs.yaml playbook (only once every 10 years, technically)

https://www.trustzone.com/ssl-certificate-validity-is-now-capped-at-a-maximum-of-2-years/
I'd stay with 1 year or so. Who knows how clients will behave for such longterm certs

**Comment by chmeliik** [comment]
Cleanup of config files (mostly removing unused options that do not seem useful in any way while keeping those that may eventually someday be relevant). For the httpd ssl config in hub, now using an up-to-date version from the httpd fedora package.

**Comment by chmeliik** [comment]
Moved common ansible vars to a file under group_vars/

**Comment by chmeliik** [comment]
Prepared variables that will eventually be used for installing our code from git

**Comment by chmeliik** [comment]
Now installing koji-containerbuild and osbs-client from the specified git repo + branch (or tag / commit)

**Comment by chmeliik** [comment]
More cleanup of ansible stuff (not using tmpdir, using a params file for OpenShift templates)

**Comment by chmeliik** [comment]
Now using actually persistent volumes (so persistent that they will need to be cleaned manually, in fact), also mounting /mnt/koji/ on the builder

**Comment by chmeliik** [comment]
Fixed the login button, also the HTTPS route works now

**Comment by chmeliik** [comment]
koji-builder _should_ new do the koji setup. Some more polish required

**Comment by chmeliik** [comment]
Koji builder setup appears to be working as intended, which means that:
* When koji-builder is deployed the first time (against a clean database), it does all of the setup as intended
* On subsequent deployments, it does the setup again, but makes sure to ignore specific errors from koji (the 'already exists' kind)
  * This means that even if the setup script changes or someone manually deletes tags / build targets / etc. the setup will be done properly (but still fail if an error is unexpected)

**Comment by chmeliik** [comment]
Updated README, decided RHEL can wait

**Comment by chmeliik** [comment]
Turns out `passthrough` routes also have the `insecureEdgeTerminationPolicy` option, so the hack to fix broken logout was unnecessary

File: client/bin/setup.sh (hunk 1 above)
Comment (chmeliik) [reply client/bin/setup.sh]: Thanks, added in `koji-builder/bin/koji-setup.sh`

**Comment by chmeliik** [comment]
Had to do it for the line count diff ¯\\\_(ツ)\_/¯

File: README.md [hunk 2]
```diff
@@ -1,137 +1,177 @@
...
 
-Inspect koji-hub container logs to view link for accessing web interface:
+* You've done some koji setup, but now you need to undo it
+* For some reason, updating your OSBS-Box failed (and it is not because of the code)
+* You are done with OSBS-Box (forever)
 
-```
-docker-compose logs koji-hub
-```
+Cleanup steps:
```
Comment (MartinBasti) [inline README.md]: Can we have cleanup playbook for that?

File: README.md [hunk 3]
```diff
@@ -1,137 +1,177 @@
...
-koji-hub_1      | + exec httpd -D FOREGROUND
-koji-hub_1      | [Thu Oct 06 15:18:37.651360 2016] [so:warn] [pid 1] AH01574: module ssl_module is already loaded, skipping
+1. Delete everything in the koji namespace: `oc delete namespace <koji namespace>`
+2. Delete koji PVs from OpenShift: `oc delete pv --selector volume=koji-volume`
+3. Delete the persistent volume directories
+    * __~/.local/share/osbs-box/pv/*__ by default
+    * They tend to be owned by other users, you may need `sudo`
+5. _OPTIONAL_: Delete all the remaining data
+    * __~/.local/share/osbs-box/__ by default
+    * You should never _need_ to do this - unless you are breaking up with OSBS-Box :cry:
```
Comment (MartinBasti) [inline README.md]: :rocket:

File: generate-certs.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      loop:
+        "{{ koji_client_certs }}"
+
+    - name: Generate PKCS#12 client certificates for browsers
+      openssl_pkcs12:
+        path: "{{ koji_certs_final }}/{{ item }}.p12"
+        privatekey_path: "{{ koji_certs_work }}/{{ item }}.key"
+        certificate_path: "{{ koji_certs_work }}/{{ item }}.crt"
+        friendly_name: "{{ item }}"
+        mode: 0644
```
Comment (MartinBasti) [inline generate-certs.yaml]: Ehm, allowing reading secret for others doesn't seem secure to me

File: koji-base/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+# Install git-core, it will be needed when pip-installing packages from git
+RUN dnf -y install "git-core" && \
+    dnf clean all
+
+COPY bin/ /usr/local/bin/
+COPY etc/ /etc/
+
+# Certificates are added in mode 0600 by openshift, but apache needs read access
+RUN chmod -R o+r /etc/pki/koji/
```
Comment (MartinBasti) [inline koji-base/Dockerfile]: Sooo, it should be readable for apache group, not for everybody?

Comment (MartinBasti) [reply koji-base/Dockerfile]: Or you can add apache into koji group?

File: koji-builder/etc/osbs.conf [hunk 6]
```diff
@@ -0,0 +1,31 @@
+[general]
+build_json_dir = /usr/share/osbs/
+verbose = 1
+openshift_required_version = 1.0.6
```
Comment (MartinBasti) [inline koji-builder/etc/osbs.conf]: I'd go for 3.6

**Pull request merged**